feat(core) : introduce is_positive_infinite and is_negative_infinite

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Introduce `is_positive_infinite` and `is_negative_infinite` assumption attributes.

#### Other comments

These attributes will be used to reorganize the facts in next PR.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
  - `.is_positive_infinite` and `.is_negative_infinite` attributes are introduced.
<!-- END RELEASE NOTES -->

JSS95 committed:
feat(core) : introduce is_positive_infinite and is_negative_infinite

JSS95 commented:
@oscarbenjamin 

Here is the highlighted benchmark result:

```
All benchmarks:

       before           after         ratio
     [85ff38c1]       [3aea815f]
     <master>         <coreassump_negposinf>
+     19.7±0.09μs       21.7±0.2μs     1.10  core.arit.Time_assumption.time_ncmul
+       175±0.7μs          194±1μs     1.11  matrices.TimeMatrixExpression.time_MatMul
+      51.5±0.9μs       57.1±0.5μs     1.11  solve.TimeMatrixArithmetic.time_dense_add(10, 0)
```

oscarbenjamin commented:
I'm guessing that the slowdown is due to `_ask` (in the old assumptions) checking the additional predicates on symbols for which they are not given any value anyway. Probably if these two new predicates can have the effect of a 10% slowdown then making assumptions checking on symbols more efficient can lead to improvements in those benchmarks. Maybe there is a way to initialise the `_assumptions` dict for a symbol with prefilled Nones e.g.:
```python
In [13]: x = Symbol('x')

In [14]: x._assumptions
Out[14]: {'commutative': True}

In [15]: x.is_positive

In [16]: x._assumptions
Out[16]: 
{'commutative': True,
 'positive': None,
 'finite': None,
 'negative': None,
 'extended_negative': None,
 'extended_positive': None}
```
Any symbol created with given assumptions will always have `None` for all of the unspecified assumptions so it's wasteful to check for `_eval_is_positive` etc.

JSS95 committed:
feat(core) : introduce is_positive_infinite and is_negative_infinite

Introduce StdFactKB.fill_None()
Make Symbol.__new_stage2__ call StdFactKB.fill_None()

JSS95 committed:
Merge branch 'master' into coreassump_negposinf

JSS95 committed:
feat(core) : introduce is_positive_infinite and is_negative_infinite

Modify StdFactKB.fill_None()

JSS95 commented:
@oscarbenjamin 
I implemented your suggestion. Now, `_assumptions` attribute of `Symbol` shows:
```python
In [1]: Symbol('x', positive=True)._assumptions                                                                               
Out[1]: 
{'positive': True,
 'negative_infinite': False,
 'nonzero': True,
 'negative': False,
 'imaginary': False,
 'extended_nonpositive': False,
 'complex': True,
 'real': True,
 'finite': True,
 'extended_positive': True,
 'extended_real': True,
 'commutative': True,
 'extended_negative': False,
 'nonpositive': False,
 'hermitian': True,
 'positive_infinite': False,
 'extended_nonzero': True,
 'infinite': False,
 'nonnegative': True,
 'extended_nonnegative': True,
 'zero': False}
```

Benchmark showed a small improvement compared to previous commit.
```
All benchmarks:

       before           after         ratio
     [3aea815f]       [373305cd]
     <coreassump_negposinf~3>       <coreassump_negposinf>
-      70.9±0.5ms       61.5±0.4ms     0.87  polygon.PolygonEnclosesPoint.time_bench01
```

JSS95 committed:
feat(core) : introduce is_positive_infinite and is_negative_infinite

Modify StdFactKB.fill_None()

JSS95 committed:
feat(core) : introduce is_positive_infinite and is_negative_infinite

Modify StdFactKB.fill_None()

JSS95 commented:
@oscarbenjamin 
I implemented your suggestion.

```python
In [1]: Symbol('x', positive=True)._assumptions                                                                               
Out[1]: 
{'positive': True,
 'infinite': False,
 'nonzero': True,
 'commutative': True,
 'negative_infinite': False,
 'extended_positive': True,
 'extended_negative': False,
 'nonnegative': True,
 'imaginary': False,
 'hermitian': True,
 'extended_nonpositive': False,
 'extended_nonzero': True,
 'real': True,
 'extended_real': True,
 'zero': False,
 'positive_infinite': False,
 'extended_nonnegative': True,
 'negative': False,
 'complex': True,
 'finite': True,
 'nonpositive': False,
 'prime': None,
 'polar': None,
 'algebraic': None,
 'integer': None,
 'transcendental': None,
 'irrational': None,
 'odd': None,
 'rational': None,
 'noninteger': None,
 'composite': None,
 'antihermitian': None,
 'even': None}
```

Benchmark showed mixed result.
```
All benchmarks:

       before           after         ratio
     [3aea815f]       [8536be87]
     <coreassump_negposinf~5>       <coreassump_negposinf>
-      77.6±0.6ms         65.3±1ms     0.84  logic.LogicSuite.time_dpll2
+         184±3μs          218±8μs     1.18  matrices.TimeMatrixExpression.time_MatMul
+      11.6±0.2ms       12.9±0.2ms     1.12  matrices.TimeMatrixExpression.time_MatMul_doit
```

oscarbenjamin commented:
It's hard to tell if any of these timings differences are actually significant.

JSS95 commented:
@oscarbenjamin 
Indeed. So, how do you think of this change? Can this be merged? Shall this be merged with or without `Symbol._assumptions` change?

oscarbenjamin commented:
I don't think we should make the change to `Symbol._assumptions` unless we can see it being advantageous in some situation. Previously something seemed to slow down which is why you broke the other PR into smaller pieces. I thought there was a slowdown in something involving matrices with symbols in but I can't remember. Is that slowdown not seen with the other changes here with or without the change to `Symbol._assumptions`?

JSS95 commented:
Yes. It seems that filling None in in `Symbol._assumptions` slightly slows down MatMul. 
```

       before           after         ratio
     [ea3aedd2]       [f47fc8eb]
     <master>         <symbol_assumptions>
+       175±0.3μs        206±0.1μs     1.18  matrices.TimeMatrixExpression.time_MatMul
+     11.1±0.02ms      12.4±0.02ms     1.11  matrices.TimeMatrixExpression.time_MatMul_doit
```

Here, `symbol_assumptions` branch is where I added the change you suggested to current master. (`is_positive/negative_infinite` attributes are not introduced there)

oscarbenjamin commented:
Maybe creating the assumptions dict for a symbol should be cached. I expect that all `MatrixSymbol` have exactly the same `_assumptions` dict. There's also probably no harm in having all `Symbol` with the same assumptions share a single instance of the dict.

JSS95 commented:
@oscarbenjamin 

Caching showed no difference.
```
       before           after         ratio
     [ea3aedd2]       [3ac2a2fa]
     <master>         <symbol_assumptions>
+         179±4μs        207±0.6μs     1.15  matrices.TimeMatrixExpression.time_MatMul
+     11.2±0.04ms      12.6±0.05ms     1.13  matrices.TimeMatrixExpression.time_MatMul_doit
```

`MatrixSymbol` is not subclass of `Symbol`. It's strange that `MatMul` shows such slowdown.

asmeurer commented:
What's the motivation for this? Isn't this the same as `is_positive and is_infinite`? We should only add "shortcut" assumptions like this if they need to be typed out a lot, but in the diff this isn't applied anywhere in library code.

asmeurer commented on sympy/core/assumptions.py:
```diff
@@ -257,6 +258,10 @@
+
+
+    'positive_infinite -> extended_positive & infinite',
+    'negative_infinite -> extended_negative & infinite',
```
Should the reverse implication also hold?

asmeurer commented on sympy/core/assumptions.py:
```diff
@@ -440,6 +445,15 @@ def copy(self):
+        all relevant facts are found by deduction.
+        """
+        for key in _assume_defined:
+            if key not in self:
```
We can also define 

```py
def __missing__(self, key):
    if key in _assume_defined:
        return None
    raise KeyError(key)
```

See https://docs.python.org/3/reference/datamodel.html#object.__missing__

asmeurer replied:
I'm not sure how this is used, though. That won't actually fill in the missing keys, so it could break code if it expects all the keys to be there. This is approach potentially more performant if potentially not every key is accessed.

JSS95 commented:
@asmeurer 
The reason to implement `is_positive/negative_infinite` (and `Q.positive/negative_infinite`) is to reorganize the fact system so that predicates such as `real` can be split into combination of primitive predicate, e.g. `negative_infinite | negative | zero | positive | positive_infinite`. This may reduce the complexity of sat solver, which grows exponentially in the number of predicates. It is suggested in [this comment](https://github.com/sympy/sympy/pull/21220#issuecomment-811758433) by @oscarbenjamin 

My ultimate goal is to introduce `Q.extended_positive` and other predicates which exist in old assumption system but do not in new system.

oscarbenjamin commented:
It isn't necessarily the case that we need to expose `is_positive_infinite` and `is_negative_infinite` in the old or new assumptions. The point was just that they could be used internally by the SAT solver so that it can canonicalise incoming predicates like `nonnegative` and `extended_real` into a smaller number of mutually exclusive predicates.

oscarbenjamin commented:
The same approach doesn't work with the old assumptions because there's no way to specify "or" in the three-valued logic e.g. it isn't possible to define a symbol that is "negative or prime".

JSS95 commented:
Oh, okay. I misinterpreted your intention. I will not introduce `is_positive/negative_infinite` to old assumptions, then.
